### ISSUE TITLE
🏗 Upon discussion: how to implement parallelized/chunked reading

### ISSUE THREAD

**Issue by philipp-baumann** [spectral-cockpit/opusreader2#23]
Currently, `read_opus_file()` reads a single file. It takes about 20 milliseconds to parse all relevant spectral blocks. This is not a lot, but still it would be nice to feature reading in parallel.

The idea of this feature is that instead of distributing single file reads across cores or other parallelized instances, we chunk/split a file list according to the registered parallel instances.

To allow maximal flexibility, we should rely on the R "future" framework. As we don't want any other dependencies than base R, there seems two suitable options for us:

1. add a `parallel_chunk` (or other suitable name) argument to `read_opus()`, the wrapper. Then, check for the availability of {future.apply} using `require()`, which stops with a graceful message if not available, and parses in parallel if the package is available.
2. Create a separate package, e.g. {readopus2prod}, which contains advanced helper utilities, and has {opusreader2} as a dependency.

**Comment by pierreroudier** [comment]
@philipp-baumann Just noting that my PR https://github.com/spectral-cockpit/opusreader2/pull/31 only implements the multiple file read sequentially, I have not added the option to add a `cluster` to `pbapply::pblapply` so several nodes can be used.

Like you, I have a feeling that the overheads would be way too high, and using chunks would be a better option. 

In my experience (as a user), a progress bar is nice when you start working with thousands of file, but of course it's not a *must* have.

**Comment by philipp-baumann** [comment]
yes. agree :-) if I have time tomorrow I can draft something tomorrow and we can further discuss.

**Comment by philipp-baumann** [comment]
progress bar is nice though. if one has to wait (as you say maybe 10000's of files, it's is nice to see how long one has to wait :)).

**Comment by philipp-baumann** [comment]
if we use future, we let the user decide how to parallelize. The developer only decides what to parallelize.

**Comment by philipp-baumann** [comment]
@ThomasKnecht , we have now `read_opus()` at top-level in the main branch. If we introduce let's say parallel chunking functionality via the `chunk_parallel` argument, we would need another top-level function. I see the following solution:

1. we rename `read_opus()` to `read_opus_impl()` 
2. we introduce a new `read_opus`, that dispatches to parallel if `chunk_parallel = TRUE`

Do you agree?

**Comment by ThomasKnecht** [comment]
see #32

**Comment by philipp-baumann** [comment]
> see #32

I like the `opus_apply()` style dispatch. Now what we need is a chunked version. Let's say we have something like this

```{R}
# register all available cores in multicore mode
library("future.apply")
plan(multiprocess) # or multicore if UNIX and a non-interactive environment
availableCores()
# then call `read_opus()` with parallel option, which then assigns to correct parallization mode
read_opus(dsn, parallel = TRUE)
```

**Comment by ThomasKnecht** [comment]
a solution is implemented here: #47

**Comment by ThomasKnecht** [comment]
can we close this?
